### PR TITLE
Add XPENDING, XINFO, XDEL, and XTRIM commands

### PR DESCRIPTION
**Pull request by smitfrank**
On the left side of the arrow is the function name from trio_redis and on the right side is the Redis command:

- `xpending` → `XPENDING key group`
- `xinfo_groups` → `XINFO GROUPS key`
- `xinfo_streams` → `XINFO STREAMS key`
- `xdel` → `XDEL key ID [ID ...]`
- `xtrim` → `XTRIM key MAXLEN [~] count`
  - **NOTE:** The `~` option, i.e. approximate trimming is on by default. Pass `approximate=False` to disable it.

Redis documentation:

- https://redis.io/commands/xpending
- https://redis.io/commands/xinfo
- https://redis.io/commands/xdel
- https://redis.io/commands/xtrim

### CODE REVIEW

File: tests/test_stream_commands.py [hunk 1]
```diff
@@ -181,3 +181,101 @@ async def test_xclaim(redis):
...
     assert result == 1
+
+
+async def test_xpending(redis):
+    key = 'x'
+    group = 'group_a'
+    consumer_1 = 'consumer_1'
+    consumer_2 = 'consumer_2'
+
+    # Prepare: create stream and group, add a message,
```
Comment (lrozema) [inline tests/test_stream_commands.py:192]:
```suggestion
    # Prepare: create stream and group, add a message
```

File: tests/test_stream_commands.py [hunk 2]
```diff
@@ -181,3 +181,101 @@ async def test_xclaim(redis):
...
+    # and read and not acknowledge message.
```
Comment (lrozema) [inline tests/test_stream_commands.py:193]:
```suggestion
    # and read and don't acknowledge message.
```

File: tests/test_stream_commands.py [hunk 3]
```diff
@@ -181,3 +181,101 @@ async def test_xclaim(redis):
...
+    key = 'x'
+    msgid = await redis.xadd(key, {b'foo': b'bar'})
+    await redis.xadd(key, {b'foo': b'bar'})
+
+    assert await redis.xlen(key) == 2
+
+    result = await redis.xdel(key, msgid)
+
+    assert result == 1
+    assert await redis.xlen(key) == 1
```
Comment (lrozema) [inline tests/test_stream_commands.py:265]: Is tested good enough? Wouldn't it be better to add a foo1/bar1 and a foo2/bar2, store msgid for both and after deleting make sure it is msgid2 that is still available (e.g. only msgid1 got deleted).

File: tests/test_stream_commands.py [hunk 4]
```diff
@@ -181,3 +181,101 @@ async def test_xclaim(redis):
...
+
+    assert result == {
+        b'first-entry': (msgid_1, payload),
+        b'groups': 1,
+        b'last-entry': (msgid_2, payload),
+        b'last-generated-id': msgid_2,
+        b'length': 2,
+        b'radix-tree-keys': 1,
+        b'radix-tree-nodes': 2,
+    }
```
Comment (lrozema) [inline tests/test_stream_commands.py:252]: Why not use different payloads for each message?

File: trio_redis/_commands.py [hunk 5]
```diff
@@ -45,12 +45,39 @@ def parse_stream_list(reply):
...
 
+def parse_xinfo_stream(response):
+    data = pairs_to_dict(response)
+    first = data[b'first-entry']
+    if first is not None:
+        data[b'first-entry'] = (first[0], pairs_to_dict(first[1]))
+    last = data[b'last-entry']
+    if last is not None:
+        data[b'last-entry'] = (last[0], pairs_to_dict(last[1]))
+    return data
```
Comment (lrozema) [inline trio_redis/_commands.py:56]:
```suggestion
def parse_xinfo_stream(response):
    data = pairs_to_dict(response)

    for k in (b'first-entry', b'last-entry'):
        item = data[k]
        if item is not None:
            data[k] = (item[0], pairs_to_dict(item[1]))

    return data
```

File: trio_redis/_commands.py [hunk 6]
```diff
@@ -45,12 +45,39 @@ def parse_stream_list(reply):
...
+        b'pending': response[0],
+        b'min': response[1],
+        b'max': response[2],
+        b'consumers': consumers
+    }
+
+
+def parse_xpending_range(reply):
+    if not reply:
+        return []
```
Comment (lrozema) [inline trio_redis/_commands.py:80]:
```suggestion
        return []

```

File: trio_redis/_commands.py [hunk 7]
```diff
@@ -201,6 +228,9 @@ def xgroup_create(self, key, groupname, id=b'$', mkstream=False):
     def xack(self, key, groupname, *ids):
         return self.execute([b'XACK', key, groupname, *ids], int)
 
+    def xpending(self, key, groupname):
+        return self.execute([b'XPENDING', key, groupname], parse_xpending)
+
     def xpending_range(self, key, groupname, start, end, count, consumername=None):
         # NOTE: The start, end, and count arguments are optional to
```
Comment (lrozema) [inline trio_redis/_commands.py:235]:
```suggestion
        # NOTE: The start, end and count arguments are optional to
```

File: trio_redis/_commands.py [hunk 8]
```diff
@@ -201,6 +228,9 @@ def xgroup_create(self, key, groupname, id=b'$', mkstream=False):
...
         # XPENDING, but we only user the command with these arguments.
```
Comment (lrozema) [inline trio_redis/_commands.py:236]:
```suggestion
        # XPENDING, but we only use the command with these arguments.
```

File: trio_redis/_commands.py [hunk 9]
```diff
@@ -250,6 +280,22 @@ def xrange(self, key, min='-', max='+', count=None):
...
+    def xinfo_stream(self, key):
+        return self.execute([b'XINFO', b'STREAM', key], parse_xinfo_stream)
+
+    def xdel(self, key, *ids):
+        return self.execute([b'XDEL', key, *ids], int)
+
+    def xtrim(self, key, maxlen, approximate=True):
+        pieces = [b'XTRIM', key, b'MAXLEN']
+        if approximate:
+            pieces.append(b'~')
```
Comment (lrozema) [inline trio_redis/_commands.py:295]:
```suggestion

        if approximate:
            pieces.append(b'~')

```

File: trio_redis/_redis.py [hunk 10]
```diff
@@ -218,6 +218,10 @@ def execute(self, command, parse_callback=None):
         return self
 
     def __await__(self):
+        if not self._buffer:
+            async def return_empty_list():
+                return []
+            return return_empty_list().__await__()
```
Comment (lrozema) [inline trio_redis/_redis.py:224]:
```suggestion
            return return_empty_list().__await__()

```

File: tests/test_stream_commands.py (hunk 3 above)
Comment (smitfrank) [reply tests/test_stream_commands.py:265]: I added asserts for the stream contents.

File: tests/test_stream_commands.py (hunk 4 above)
Comment (smitfrank) [reply tests/test_stream_commands.py]: Done.

File: trio_redis/testing_utils.py [hunk 11]
```diff
@@ -193,3 +194,51 @@ async def fail_if_not_between(after, before, coroutine):
...
+    async def handle_client(self, client_stream):
+        target_stream = await trio.open_tcp_stream(self.host, self.port)
+
+        try:
+            while True:
+                # NOTE: We're assuming the requests from then client and
+                # the responses from the server are never bigger than
+                # 64KiB (Trio's default). This assumption makes the code
+                # below much simpler. There's no logic needed to detect
+                # if a request or response it complete.
```
Comment (lrozema) [inline trio_redis/testing_utils.py:225]:
```suggestion
                # if a request or response is complete.
```

File: trio_redis/testing_utils.py [hunk 12]
```diff
@@ -193,3 +194,51 @@ async def fail_if_not_between(after, before, coroutine):
...
+
+    def _get_address_from_listeners(self, listeners):
+        return listeners[0].socket.getsockname()
+
+    async def handle_client(self, client_stream):
+        target_stream = await trio.open_tcp_stream(self.host, self.port)
+
+        try:
+            while True:
+                # NOTE: We're assuming the requests from then client and
```
Comment (lrozema) [inline trio_redis/testing_utils.py:221]:
```suggestion
                # NOTE: We're assuming the requests from the client and
```

File: trio_redis/testing_utils.py [hunk 13]
```diff
@@ -193,3 +194,51 @@ async def fail_if_not_between(after, before, coroutine):
...
+                # the responses from the server are never bigger than
+                # 64KiB (Trio's default). This assumption makes the code
+                # below much simpler. There's no logic needed to detect
+                # if a request or response it complete.
+                request = await client_stream.receive_some()
+                request = await self.handle_request(request)
+                await target_stream.send_all(request)
+                response = await target_stream.receive_some()
```
Comment (lrozema) [inline trio_redis/testing_utils.py:229]:
```suggestion

                response = await target_stream.receive_some()
```

**Pull request merged**